ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -159,48 +161,66 @@
         break;
       case UndeleteOperation:
-        diskParallelism = routerConfig.routerUndeleteRequestParallelism;
+        diskReplicaParallelism = routerConfig.routerUndeleteRequestParallelism;
         crossColoEnabled = true;
         eligibleReplicas = getEligibleReplicas(partitionId, null,
             EnumSet.of(ReplicaState.BOOTSTRAP, ReplicaState.STANDBY, ReplicaState.LEADER));
         // Undelete operation need to get global quorum. It will require a different criteria for success.
         // Here set the success target to the number of eligible replicas.
-        diskSuccessTarget = eligibleReplicas.size();
+        diskReplicaSuccessTarget = eligibleReplicas.size();
         break;
       default:
         throw new IllegalArgumentException("Unsupported operation: " + routerOperation);
     }
-    if (diskParallelism < 1 || cloudParallelism < 1) {
+    if (diskReplicaParallelism < 1 || cloudReplicaParallelism < 1) {
       throw new IllegalArgumentException(
-          "Parallelism has to be > 0. diskParallelism=" + diskParallelism + ", cloudParallelism=" + cloudParallelism
-              + ", routerOperation=" + routerOperation);
+          "Parallelism has to be > 0. diskParallelism=" + diskReplicaParallelism + ", cloudParallelism="
+              + cloudReplicaParallelism + ", routerOperation=" + routerOperation);
     }
 
     // Order the replicas so that local healthy replicas are ordered and returned first,
     // then the remote healthy ones, and finally the possibly down ones.
     List<? extends ReplicaId> replicas =
         routerConfig.routerGetEligibleReplicasByStateEnabled ? eligibleReplicas : partitionId.getReplicaIds();
+    // In a case where a certain dc is decommissioned and blobs previously uploaded to this dc now have a unrecognizable
+    // dc id. Current clustermap code will treat originating dc as null if dc id is not identifiable. To improve success
+    // rate of cross-colo requests(GET/DELETE/TTLUpdate), operation tracker should be allowed to try remote dc with most
+    // replicas first. This is useful in cluster with "unbalanced" replica distribution (i.e. 3 replicas in local dc and
+    // 1 replica per remote dc)
+    if (originatingDcName == null && routerConfig.routerCrossColoRequestToDcWithMostReplicas) {
+      Map<String, Long> dcToReplicaCnt = replicas.stream()
+          .collect(Collectors.groupingBy(e -> e.getDataNodeId().getDatacenterName(), Collectors.counting()));
+      List<Map.Entry<String, Long>> entryList = new ArrayList<>(dcToReplicaCnt.entrySet());
+      entryList.sort(Map.Entry.comparingByValue());
+      // we assign a dc with most replicas to "originatingDcName", which only takes effect when populating replica pool
+      // (replicas in that colo have higher priority than other remote colos). Note that, "this.originatingDcName" still
+      // keeps the actual originating dc name (which is null). This value forces operation track to go through replicas
```

Comment:
Minor: operation tracker

Reply:
I recommend using a different variable name than originatingDcName.  Making it different from the class variable is too confusing.  Also, I don't see where local originatingDcName is being used other than in the debug line.